### ISSUE TITLE
[HOLD for payment 2023-12-12] [$500] Distance – Inconsistency when editing waypoints

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#30290]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.90.1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @paultsimura
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1698158967101059

## Action Performed:
1. Click FAB - Request Money - Distance
2. Fill in Start and Finish points
3. Create the request
4. Navigate to the created request
5. Click "distance" to open "Edit distance" modal
6. Add a waypoint and click "save"
7. Click "distance" again and remove any of the waypoints
8. Click "Save"
9. Click "Amount" and modify the amount
10. Reload the page or switch to another chat and back

## Expected Result:
 After the amount is modified manually, it should persist.

## Actual Result:
The modified amount is reverted to the previous value

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

 

https://github.com/Expensify/App/assets/93399543/c2629d7e-6564-4679-a403-8c984b484e20

</details>

<details>
<summary>iOS: Native</summary>

 

https://github.com/Expensify/App/assets/93399543/8cef06e7-275c-4ce5-a9d2-0af6a6aa7b35

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/93399543/c62baeb3-8d98-4a97-ac7a-a87feafafd44

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/93399543/715b8150-4ae0-4a31-b766-77cf28286686

https://github.com/Expensify/App/assets/93399543/ac25ed16-7e0a-4eb2-bf6b-86ecda414c9e

</details>

<details>
<summary>MacOS: Desktop</summary>

 

https://github.com/Expensify/App/assets/93399543/353790dc-0604-4975-9b60-72f1081e8f51

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014a1428f6ef24d6ec</li>
        <li>Upwork Job ID: 1716905913390137344</li>
        <li>Last Price Increase: 2023-11-07</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Contributor | 27653898</li>
<li>paultsimura | Reporter | 27653901</li>
</ul></li>
    </ul>
</details>

**Comment by kbecciv** [comment]
Proposal by: @paultsimura
Slack conversation: https://expensify.slack.com/archives/C049HHMV9SM/p1698158967101059

## Proposal
### Please re-state the problem that we are trying to solve in this issue.
When the distance request is modified: added, removed waypoint, and manually changed amount, it's reverted, and the removed waypoint is back.
### What is the root cause of that problem?
When executing the `UpdateDistanceRequest` API call, we build the Onyx data with waypoints of the updated transaction here:
https://github.com/Expensify/App/blob/b1ae0c20bd6926cf03ebd7d011beb2ee61c1814a/src/libs/TransactionUtils.ts#L148-L151
Let's take a look at what happens when we remove one waypoint.
The simple merge operation for the following object:
```js
{
    "comment": {
        "type": "customUnit",
        "customUnit": {
            "name": "Distance"
        },
        "waypoints": {
            "waypoint0": { wp-0 },
            "waypoint1": { wp-1 },
            "waypoint2": { wp-2 }
        }
    }
}
```
To the following object:
```js
{
    "comment": {
        "type": "customUnit",
        "customUnit": {
            "name": "Distance"
        },
        "waypoints": {
            "waypoint0": { wp-0 },
            "waypoint1": { wp-1 },
        }
    }
}
```
...will result in the same as pre-merge:
```js
{
    "comment": {
        "type": "customUnit",
        "customUnit": {
            "name": "Distance"
        },
        "waypoints": {
            "waypoint0": { wp-0 },
            "waypoint1": { wp-1 },
            "waypoint2": { wp-2 }
        }
    }
}
```
This is because the `waypoint2` merge key was not passed to Onyx, so it's ignored, as Onyx replaces values key-by-key to the latest nested object. So this merge operation modifies only `waypoint0` and `waypoint1`, leaving the `waypoint2` untouched.
These waypoints are later used to build the next optimistic transaction when modifying the request, which causes the buggy behavior.
### What changes do you think we should make in order to solve the problem?
We should explicitly mark the deleted waypoints as `null` when passing them to Onyx.
**First**, create a new function in `TransactionUtils`:
```ts
function getReplacementWaypoints(transaction: Transaction, updatedWaypoints: WaypointCollection): WaypointCollection {
    const existingWaypoints = getWaypoints(transaction) ?? {};
    if (Object.keys(updatedWaypoints).length >= Object.keys(existingWaypoints).length) {
        return updatedWaypoints;
    }
    return lodashReduce(existingWaypoints, (result, value, key) => ({
        ...result,
        // Explicitly set the non-existing waypoints to null so that Onyx can remove them
        [key]: updatedWaypoints[key] ?? null,
    }), {});
}
```
**Second**, use it here when building the updated transaction:
```diff
    if (Object.hasOwn(transactionChanges, 'waypoints')) {
-       updatedTransaction.modifiedWaypoints = transactionChanges.waypoints;
+       updatedTransaction.modifiedWaypoints = getReplacementWaypoints(updatedTransaction, transactionChanges.waypoints!!);
        shouldStopSmartscan = true;
    }
```
**Third**, we are also passing these waypoints to MapBox on the API call. We pass them as a stringified JSON, and there we should remove the null-ish waypoints:
https://github.com/Expensify/App/blob/3b86da6dc1ec2ca15033bbdc09702892966c2580/src/libs/actions/IOU.js#L697
We should use the following:
```js
waypoints: JSON.stringify(TransactionUtils.getValidWaypoints(transactionDetails.waypoints, true)),
```
Similar to what we do in the `createDistanceRequest`:
https://github.com/Expensify/App/blob/3b86da6dc1ec2ca15033bbdc09702892966c2580/src/libs/actions/IOU.js#L649
#### Result:
https://github.com/Expensify/App/assets/12595293/cbf9d9a0-f337-4032-b1c8-ea12e109e4e2
### What alternative solutions did you explore? (Optional)
Alternatively, we can only modify the `Transaction.removeWaypoint` function to set not only `waypoints`, but also the `modifiedWaypoints`:
```ts
    let newTransaction: Transaction = {
        ...transaction,
        comment: {
            ...transaction.comment,
            waypoints: reIndexedWaypoints,
        },
        modifiedWaypoints: reIndexedWaypoints
    };
```
https://github.com/Expensify/App/blob/01016dcaf4da744fe554cc839b47b5780f01c637/src/libs/actions/Transaction.ts#L125-L134
This solution is simpler and will have the same effect, but I'm not sure if we are OK with modifying the `modifiedWaypoints` field before the actual API call – would appreciate C+ input on this.

**Comment by DylanDylann** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Distance – Inconsistency when editing waypoints

### What is the root cause of that problem?
For example, on ONYX we have:
```
waypoints: {
   "waypoint1": {before},
   "waypoint2": {before},
   "waypoint3:" {before }
}
```

When remove a waypoint we will update waypoints field like this
method: Merge
optimisticData:
```
waypoints: {
   "waypoint1": {after},
   "waypoint2": {after},
}
```

As how ONYX.merge work, the result will be 
```
waypoints: {
   "waypoint1": {after},
   "waypoint2": {after},
   "waypoint3:" {before }.  //**still remain**
}
```

This problem happens with transaction.modifiedWaypoints that causes this bug

### What changes do you think we should make in order to solve the problem?
Note that: the same problem doesn't happen with transaction.comment.waypoint field because we are using SET method instead of MERGE method
https://github.com/Expensify/App/blob/cd3c863bcc26ec84d27e4dd485e676af1359f5f2/src/libs/actions/Transaction.ts#L154

So I suggest 2 solutions for this bug:
1. When removing the waypoint we will use SET method instead of MERGE method like we did here https://github.com/Expensify/App/blob/cd3c863bcc26ec84d27e4dd485e676af1359f5f2/src/libs/actions/Transaction.ts#L154
2. When removing the waypoint we reset waypoint to null if transactionChanges.waypoints is not exist
in here
https://github.com/Expensify/App/blob/a8482e20e92916af993bf3198036dbd0f412f9dd/src/libs/TransactionUtils.ts#L149
update like this
```
const waypointKeys = Object.keys(transaction.modifiedWaypoints ?? {})
        const reIndexedWaypoints: WaypointCollection = {};
        waypointKeys.forEach((waypoint, index) => {
            reIndexedWaypoints[`waypoint${index}`] = transactionChanges.waypoints && transactionChanges.waypoints[`waypoint${index}`] ? transactionChanges.waypoints[`waypoint${index}`] : null
        });
       
        updatedTransaction.modifiedWaypoints = reIndexedWaypoints;
```
Then using getValidWaypoints function to remove null value when calling API
### What alternative solutions did you explore? (Optional)

**Comment by paultsimura** [comment]
@DylanDylann [that's](https://github.com/Expensify/App/issues/30290#issuecomment-1778957070) basically 1:1 [my proposal](https://github.com/Expensify/App/issues/30290#issuecomment-1777934203), just rephrased, isn't it?

**Comment by DylanDylann** [comment]
@paultsimura There are many different between 2 proposal:
1. My first solution: using SET method instead of MERGE method (you didn't mentioend, right?)
2. In the second solution I suggest to write some code line instead of make new function (I think It is clean and clearer) and suggest to make code consistency

**Comment by paultsimura** [comment]
> 1. My first solution: using SET method instead of MERGE method (you didn't mentioend, right?)

This is exactly what I suggest as an alternative solution 🙂 
Anyway, I only wanted to understand the difference, maybe I was missing something. The further is up to C+

**Comment by DylanDylann** [comment]
@paultsimura in your alternative solution, you suggest update modifiedWaypoints in removeWayPoint function. My proposal suggest using SET method instead of MERGE method for updated transaction in optimistic data. It is different totally

**Comment by jliexpensify** [comment]
Bump @Santhosh-Sellavel for reviews please!

**Comment by jliexpensify** [comment]
Bump @Santhosh-Sellavel for reviews!

**Comment by Santhosh-Sellavel** [comment]
On my list today

**Comment by Santhosh-Sellavel** [comment]
Did some investigation today, still need to dig more. Will post an review update tomorrow!

**Comment by jliexpensify** [comment]
@Santhosh-Sellavel any updates?

**Comment by Santhosh-Sellavel** [comment]
Sorry no update yet. Will post update before monday

**Comment by Santhosh-Sellavel** [comment]
Alright, I couldn't find more time for this, it's better to assign a new C+  to move this forward. cc: @jliexpensify.

I'm unassigning myself, sorry for the delay.

**Comment by jliexpensify** [comment]
Hi @rushatgabhane - could you take over and review the proposals? Thanks!

**Comment by rushatgabhane** [comment]
sure thing

**Comment by rushatgabhane** [comment]
is this still reproducible?

**Comment by paultsimura** [comment]
@rushatgabhane yes, I see on your screenshot that your 3rd change was a distance change, while it should be a manual amount change.

Just reproduced:

https://github.com/Expensify/App/assets/12595293/ecb208a0-5f52-4971-b428-8045674229a9

**Comment by rushatgabhane** [comment]
Oh ok! It's reproducible. I must have missed a step in previous try.

<img width="650" alt="image" src="https://github.com/Expensify/App/assets/29673073/f4332144-9be0-4319-b73b-77195b728381">

**Comment by rushatgabhane** [comment]
@paultsimura was first with the correct root cause and a OK [solution](https://github.com/Expensify/App/issues/30290#issuecomment-1777934203). 

But I prefer @DylanDylann's first [solution](https://github.com/Expensify/App/issues/30290#issuecomment-1778957070) to use `Onyx.Set()` 

🎀 👀 🎀

**Comment by rushatgabhane** [comment]
@techievivek im not sure who we should assign. But I think we should use the `Onyx.Set()` solution regardless. What do you think?

**Comment by paultsimura** [comment]
@rushatgabhane in my alternative solution, I already suggest doing the following:

> Alternatively, we can only modify the Transaction.removeWaypoint function to set not only waypoints, but also the modifiedWaypoints:

    let newTransaction: Transaction = {
        ...transaction,
        comment: {
            ...transaction.comment,
            waypoints: reIndexedWaypoints,
        },
        modifiedWaypoints: reIndexedWaypoints
    };


This will result in calling the `Onyx.Set()` with updated waypoints here, since this code is inside `Transaction.removeWaypoint`:

https://github.com/Expensify/App/blob/2f4b98a3b4a2700bb9b1d4daba4aff369b8db9a9/src/libs/actions/Transaction.ts#L158

IMO, this behavior is the same as calling `Onyx.Set()` elsewhere. Please correct me if I'm wrong.

**Comment by techievivek** [comment]
@rushatgabhane Happy Diwali Rushat. This is a tricky situation 😄  I will reach out to @neil-marcellini to help me select the proposal here since he would have context on the changes.

**Comment by neil-marcellini** [comment]
> @paultsimura was first with the correct root cause and a OK [solution](https://github.com/Expensify/App/issues/30290#issuecomment-1777934203).
> 
> But I prefer @DylanDylann's first [solution](https://github.com/Expensify/App/issues/30290#issuecomment-1778957070) to use `Onyx.Set()`
> 
> 🎀 👀 🎀

I agree. @DylanDylann's solution is a bit more simple. Hired.

**Comment by paultsimura** [comment]
@neil-marcellini just to confirm, did you see [this my comment](https://github.com/Expensify/App/issues/30290#issuecomment-1807573223)? Didn't it make sense for you in terms of using the `Onyx.set()`?

**Comment by DylanDylann** [comment]
I am working on PR

**Comment by DylanDylann** [comment]
@rushatgabhane The PR is ready for review https://github.com/Expensify/App/pull/31340

**Comment by paultsimura** [comment]
Hi @rushatgabhane @DylanDylann, since our C+ is currently unavailable, I'll have to inform you that this PR has introduced a regression: https://github.com/Expensify/App/issues/32511

Please take a look (maybe you will disagree) as this issue is still on the regression period.

**Comment by DylanDylann** [comment]
@paultsimura It is not a regression. Please check carefully

**Comment by jliexpensify** [comment]
Bump @rushatgabhane to complete the checklist.

@DylanDylann @paultsimura it looks like we are waiting on Manan to determine if there's a regression correct?

**Comment by paultsimura** [comment]
> looks like we are waiting on Manan to determine if there's a regression, correct?

Yes, I would say so

**Comment by jliexpensify** [comment]
Hmm so it looks like @situchan got assigned to https://github.com/Expensify/App/issues/32511 - is he able to confirm if this is a regression?

**Comment by rushatgabhane** [comment]
Hmmmm I doubt #32511 is a regression. Doesn't look related to changes made in PR https://github.com/Expensify/App/pull/31340

**Comment by paultsimura** [comment]
> Doesn't look related to changes made in PR

Let me explain my point of view.
PR https://github.com/Expensify/App/pull/31340 introduced quite a risky change: using `SET` instead of `MERGE`. Which fully replaces an object in Onyx instead of updating it field-by-field. It was said in a couple of places (e.g. here https://github.com/Expensify/App/pull/31340#pullrequestreview-1756083111) that it's quite a risky change but we should be ok considering that we **correctly build an updated transaction** object.

Therefore, when implementing such a change, the contributor must put double effort into verifying that we indeed are building a fully updated object before replacing it in Onyx. This wasn't done: the `pendingFields` of the updated transaction are not building properly, which causes https://github.com/Expensify/App/issues/32511.

Moreover, changing `SET` back to `MERGE` (i.e. reverting https://github.com/Expensify/App/pull/31340) solves https://github.com/Expensify/App/issues/32511. The issue did not happen when using `MERGE` because we updated the transaction field-by-field, and it was just appending new `pendingFields` instead of fully replacing them now.

Considering all of the above, technically it should be considered a regression, even though it was not an intentional change. But that's why it's called a regression, isn't it?

Please correct me where I'm wrong @rushatgabhane @situchan

**Comment by jliexpensify** [comment]
@rushatgabhane @situchan what's the verdict here? @neil-marcellini - is there a need for you to also weigh in?

**Comment by bondydaa** [comment]
brought up a discussion here https://expensify.slack.com/archives/C01GTK53T8Q/p1702656464396679 about potentially reverting https://github.com/Expensify/App/pull/31340 since it's caused a couple of blockers, please chime in there.

**Comment by bondydaa** [comment]
okay to recap the thread, we are going to revert https://github.com/Expensify/App/pull/31340.

Tim and Neil are still discussing alternative solutions to solve this GH though in the thread so feel free to chime in there.

**Comment by jliexpensify** [comment]
Thanks Bondy! I'll hold on the conclusion of that thread, but I'm also OOO from the 19th - 28th so will handle payments after then.

**Comment by rushatgabhane** [comment]
> It was said in a couple of places (e.g. here https://github.com/Expensify/App/pull/31340#pullrequestreview-1756083111) that it's quite a risky change but we should be ok considering that we correctly build an updated transaction object

@paultsimura thank you for the detailed explaination

**Comment by paultsimura** [comment]
According to [this thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1702911081491879?thread_ts=1702656464.396679&cid=C01GTK53T8Q), we should put this issue on hold until https://github.com/Expensify/App/issues/28358 is done

**Comment by paultsimura** [comment]
> Still on hold, right @paultsimura ?

Yes, the big 1:1:1 refactoring is currently in active development mode

**Comment by neil-marcellini** [comment]
The refactor completed 3 days ago, so we can take this off hold finally

**Comment by jliexpensify** [comment]
Wow, old issue! Are @DylanDylann and @rushatgabhane eligible for payment here?

**Comment by paultsimura** [comment]
> Wow, old issue! Are @DylanDylann and @rushatgabhane eligible for payment here?

No, the solution here was reverted as ot caused multiple regressions. After the revert, the issue was held for the 1:1:1 refactor for a second try.

**Comment by jliexpensify** [comment]
Closing this as no payments required.
